mmdet/core/hook/wandblogger_hook.py

```diff
@@ -0,0 +1,443 @@
+                    if self.log_checkpoint_metadata and self.eval_hook:
+                        metadata = self._get_ckpt_metadata(runner)
+                        aliases = [f'epoch_{runner.epoch+1}', 'latest']
+                        if self._is_best_ckpt(metadata):
+                            aliases.append('best')
+                        self._log_ckpt_as_artifact(self.ckpt_hook.out_dir,
+                                                   runner.epoch, aliases,
+                                                   metadata)
+                    else:
+                        aliases = [f'epoch_{runner.epoch+1}', 'latest']
+                        self._log_ckpt_as_artifact(self.ckpt_hook.out_dir,
+                                                   runner.epoch, aliases)
+
+        if self.num_eval_images > 0 and self.log_evaluation:
+            if self.eval_hook.by_epoch and self.eval_hook._should_evaluate(
+                    runner):
+                results = self.eval_hook.results
+                # Initialize evaluation table
+                self._init_pred_table()
+                # Log predictions
+                self._log_predictions(results, runner.epoch + 1)
+                # Log the table
+                self._log_eval_table()
+
+    @master_only
+    def after_run(self, runner):
+        self.wandb.finish()
+
+    def _log_ckpt_as_artifact(self,
+                              path_to_model,
+                              epoch,
+                              aliases,
+                              metadata=None):
+        """Log model checkpoint as  W&B Artifact.
+
+        Args:
+            path_to_model (str): Path where model checkpoints are saved.
+            epoch (int): The current epoch.
+            aliases (list): List of the aliases associated with this artifact.
+            metadata (dict, optional): Metadata associated with this artifact.
+        """
+        model_artifact = self.wandb.Artifact(
+            f'run_{self.wandb.run.id}_model', type='model', metadata=metadata)
+        model_artifact.add_file(f'{path_to_model}/epoch_{epoch+1}.pth')
+        self.wandb.log_artifact(model_artifact, aliases=aliases)
+
+    def _get_ckpt_metadata(self, runner):
+        """Get model checkpoint metadata."""
+        if self.ckpt_hook.interval == self.eval_hook.interval:
+            results = self.eval_hook.results
+        else:
+            runner.logger.info(
+                f'Evaluating for model checkpoint at epoch '
+                f'{runner.epoch+1} which will be saved as W&B Artifact.')
+            if isinstance(self.eval_hook, EvalHook):
+                from mmdet.apis import single_gpu_test
+                results = single_gpu_test(
+                    runner.model, self.val_dataloader, show=False)
+            elif isinstance(self.eval_hook, DistEvalHook):
+                from mmdet.apis import multi_gpu_test
+                results = multi_gpu_test(
```

Comment:
It is unreasonable to need an evaluation in order to get meta. If there is no other way, I think it's better not to support this feature, what do you think?

Reply:
Are you are saying that it's not reasonable to run an extra evaluation step just to get metadata for checkpoint? 

To be honest I was sceptical as well since it will be a computational overhead. This is more of an issue especially if the dataset is huge. However, this extra evaluation will happen at each checkpoint interval and ideally for longer training checkpointing is done at reasonable intervals. 

I think it's best to use modulo operator with checkpoint interval and eval interval. If the modulo is zero then we have the `results` and need not evaluate again. Bottomline is it's best if we remove this.

Reply:
Yes, In principle,  We can't have an additional evaluation process. If there is no current evaluation, consider skipping it.